### ISSUE TITLE
Warning when fine-tuning scGPT_Human with use_flash_attn=true: Missing/Unexpected keys in state_dict

### ISSUE THREAD

**Issue by wolfQK** [RCHENLAB/scGPT_fineTune_protocol#2]
Hello,  thank you for maintaining this project!

When I try to fine-tune the scGPT_Human model with the parameter `use_flash_attn` set to `True`, I receive the following warning:

> scGPT - INFO - Create new model instance ...
> scGPT - WARNING - ⚠️  Unable to load all parameters at once: Error(s) in loading state_dict for TransformerModel:
> 	Missing key(s) in state_dict: "cls_decoder._decoder.0.weight", "cls_decoder._decoder.0.bias", "cls_decoder._decoder.2.weight", "cls_decoder._decoder.2.bias", "cls_decoder._decoder.3.weight", "cls_decoder._decoder.3.bias", "cls_decoder._decoder.5.weight", "cls_decoder._decoder.5.bias", "cls_decoder._decoder.6.weight", "cls_decoder._decoder.6.bias", "cls_decoder._decoder.8.weight", "cls_decoder._decoder.8.bias", "cls_decoder._decoder.9.weight", "cls_decoder._decoder.9.bias", "cls_decoder._decoder.11.weight", "cls_decoder._decoder.11.bias", "cls_decoder._decoder.12.weight", "cls_decoder._decoder.12.bias", "cls_decoder._decoder.14.weight", "cls_decoder._decoder.14.bias", "cls_decoder._decoder.15.weight", "cls_decoder._decoder.15.bias", "cls_decoder._decoder.17.weight", "cls_decoder._decoder.17.bias", "cls_decoder.out_layer.weight", "cls_decoder.out_layer.bias". 
> 	Unexpected key(s) in state_dict: "flag_encoder.weight", "mvc_decoder.gene2query.weight", "mvc_decoder.gene2query.bias", "mvc_decoder.W.weight". 
> scGPT - INFO -    Attempting to partially load matching parameters...
> scGPT - INFO -    Successfully loaded 159/185 parameters
> scGPT - INFO - Total Pre freeze Params 52397078
> scGPT - INFO - Total Post freeze Params 52397078
> scGPT - INFO - Define loss metrics and optimizer ...
> scGPT - INFO - Start training ...

Comparing logs, I noticed in my run :

> Total Pre freeze Params 52,397,078
> Total Post freeze Params 52,397,078

In your examples:
> Total Pre freeze Params 52,443,248
> Total Post freeze Params 52,443,248

There is small difference in the parameter counts the two runs.

My Questions:
1. Could you clarify if this difference in total parameter counts my run and your logs will affect the-tuning of model?
2. Will the missing and unexpected reported in the warning have any impact on training quality or the final performance of the fine-tuned model?
3. Is this behavior expected when use_flash_attn is enabled, or does it a mismatch the model weights and loaded architecture?

Thank you very much for your time and for maintaining this project.

@Ding3LI @RuiLuo111 @RCHENLAB

**Comment by wolfQK** [comment]
Hi @Ding3LI @RuiLuo111 @RCHENLAB,

Following up on my previous issue regarding the warnings that appeared when using use_flash_attn=True during fine-tuning — I have now completed the fine-tuning process successfully with the following metrics:

> valid loss/mse: 0.0562 | err: 0.0136

However, when I used this fine-tuned model for cell type annotation, the results turned out to be unexpectedly poor.
During model loading, I again encountered a warning about unmatched parameters:

> scGPT - INFO - Create new model instance ...
> scGPT - INFO - Reading id2type.json from save_dir: /hwstorage/oe-scrna/fengqikai/debug/celltype-annotation/scGPT_test/results/annotate/Colon/HT2019-6945/Colon_id2type.json
> scGPT - WARNING - ⚠️  Unable to load all parameters at once:
> Error(s) in loading state_dict for TransformerModel:
>     Unexpected key(s) in state_dict:
>         "cls_decoder._decoder.6.weight", "cls_decoder._decoder.6.bias", 
>         "cls_decoder._decoder.8.weight", "cls_decoder._decoder.8.bias",
>         "cls_decoder._decoder.9.weight", "cls_decoder._decoder.9.bias", 
>         "cls_decoder._decoder.11.weight", "cls_decoder._decoder.11.bias",
>         "cls_decoder._decoder.12.weight", "cls_decoder._decoder.12.bias",
>         "cls_decoder._decoder.14.weight", "cls_decoder._decoder.14.bias",
>         "cls_decoder._decoder.15.weight", "cls_decoder._decoder.15.bias",
>         "cls_decoder._decoder.17.weight", "cls_decoder._decoder.17.bias".
> scGPT - INFO - Attempting to partially load matching parameters...
> scGPT - INFO - Successfully loaded 169/169 parameters

After this, the annotation results were extremely poor:

> Overall accuracy below 5%
> Confidence scores below 0.3 

This makes me wonder whether these loading issues (or the structural differences introduced when enabling use_flash_attn=True) might have affected the fine-tuning process or the model compatibility when used for cell type annotation.

Could you please advise on:

1. Whether the parameter mismatches and warnings might have degraded the fine-tuned model’s performance for downstream tasks;
2. If using use_flash_attn=False is recommended in such cases to ensure model compatibility;
3. Any specific steps I can take to check whether the fine-tuned weights were properly applied during annotation.

Thank you very much for your time and kind help!

**Comment by Ding3LI** [comment]
Hi @wolfQK , you are on the right direction.

**Short answers:**
1. **_Yes_**, parameter mismatches (or missing weights) can degrade the performance.
2. **_Yes_**, I recommend turning off the flash-attn module.
3. If you turn off flash-attn and the pre-trained model weights load successfully, with no logs indicating “Error(s) in loading state_dict...”, this means the weights are compatible.

More specifically, the main reason for turning off flash-attn is that scGPT was trained using flash-attn version 1, which has been deprecated for a long time. The attention block implementation in that version is substantially different from newer versions. I previously attempted to use flash-attn version 2 with scGPT by updating the attention code, but as you observed, some weights can still be missing because the attention block construction is different so that the performance drops dramatically.

Therefore, please proceed by turning off flash-attn and using the conventional attention blocks originally implemented in scGPT. I have also attached a Dockerfile that builds the computing environment for this protocol, in case it is helpful. Thank you for your questions.

```bash
# Use NVIDIA CUDA base image with Ubuntu
FROM nvidia/cuda:12.4.0-devel-ubuntu22.04

# Set environment variables
ENV TZ=America/Chicago \
    DEBIAN_FRONTEND=noninteractive \
    PYTHONUNBUFFERED=1 \
    PYTHONDONTWRITEBYTECODE=1 \
    CUDA_DEVICE_ORDER=PCI_BUS_ID \
    NVIDIA_VISIBLE_DEVICES=all \
    NVIDIA_DRIVER_CAPABILITIES=compute,utility \
    nnUNet_n_proc_DA=32

# Set up the working directory
WORKDIR /app

# Install system dependencies
RUN apt-get update && apt-get install -y --no-install-recommends \
    python3 python3-pip python3-dev \
    libgl1 graphviz build-essential tzdata \
    && rm -rf /var/lib/apt/lists/*

# Set timezone
RUN ln -fs /usr/share/zoneinfo/America/Chicago /etc/localtime && dpkg-reconfigure -f noninteractive tzdata

RUN ln -s /usr/bin/python3 /usr/bin/python

# Upgrade pip, setuptools, and wheel
RUN pip3 install --no-cache-dir --upgrade pip setuptools wheel

# Install PyTorch with CUDA 12.4
RUN pip3 install torch torchvision torchaudio --index-url https://download.pytorch.org/whl/cu124

# Install required Python packages
RUN pip3 install --no-cache-dir \
    scanpy scvi-tools wandb nvitop \
    louvain tqdm pyyaml \
    ipython "llvmlite>=0.38.0" \
    scgpt numpy==1.26.4 scipy==1.13.1

# Install flash-attn with no build isolation
# RUN pip3 install --no-build-isolation flash-attn==1.0.5

CMD ["/bin/bash"]
```

**Comment by wolfQK** [comment]
Hi~ @Ding3LI

Thank you very much for taking the time to look into my questions and provide such a detailed and helpful reply.

I have one more question:
If I would like to train an scGPT model from scratch for another species, for example, to build models for different mouse organs, and I want to enable flash-att for acceleration, what would be the recommended way to do this?

Could you please share any tutorials, documentation or example code related to this setup?

Thank you again your time and kind help!

**Comment by Ding3LI** [comment]
@wolfQK 

> If I would like to train an scGPT model from scratch for another species, for example, to build models for different mouse organs, and I want to enable flash-att for acceleration, what would be the recommended way to do this?

It depends on 2 situations:
1.If you truly mean to train scGPT from scratch without using any pre-trained weights, then it should be **_acceptable_** to use flash-attn version 2. The reason is that scGPT itself does not implement a customized multi-head attention module, so it can directly apply the flash-attn v1 MHA replacement in the codebase. Therefore, switching to flash-attn v2 mainly replaces the attention blocks with a more efficient implementation, leading to higher performance in terms of speed and lower memory overhead. One important point to be aware of is that training such a large foundation model with a large number of parameters requires a substantial amount of data to properly converge and learn meaningful weights.
2. If you plan to use pre-trained weights and then continue training on your own dataset, I do **_NOT_** recommend switching to flash-attn. The reason is that the released pre-trained model weights were trained assuming flash-attn v1 or conventional attention blocks, and changing the attention implementation may introduce numerical or behavioral mismatches that could negatively affect training stability or performance.

So far, I do not aware of any tutorials using flash-attn2 to train the scGPT from scratch, but I am sure that you can find some references online. I hope this is helpful for your research. Look forward to your successful training.

**Comment by wolfQK** [comment]
Hi @Ding3LI 
Thank you so much for your detailed reply — it’s extremely helpful and has clarified my doubts using flash-attn!  
I sincerely appreciate your help and look forward to further discussions with you.
